src/ray/gcs/redis_async_context.cc

```diff
@@ -109,6 +118,98 @@ Status RedisAsyncContext::RedisAsyncCommandArgv(redisCallbackFn *fn,
   return Status::OK();
 }
 
-}  // namespace gcs
+void RedisAsyncContext::Operate() {
+  if (read_requested_ && !read_in_progress_) {
+    read_in_progress_ = true;
+    socket_.async_read_some(
+        boost::asio::null_buffers(),
+        boost::bind(
+            &RedisAsyncContext::HandleIo, this, boost::asio::placeholders::error, false));
+  }
+
+  if (write_requested_ && !write_in_progress_) {
+    write_in_progress_ = true;
+    socket_.async_write_some(
+        boost::asio::null_buffers(),
+        boost::bind(
+            &RedisAsyncContext::HandleIo, this, boost::asio::placeholders::error, true));
+  }
+}
+
+void RedisAsyncContext::HandleIo(boost::system::error_code error_code, bool write) {
+  RAY_CHECK(!error_code || error_code == boost::asio::error::would_block ||
+            error_code == boost::asio::error::connection_reset ||
+            error_code == boost::asio::error::operation_aborted)
+      << "handle_io(error_code = " << error_code << ")";
+  (write ? write_in_progress_ : read_in_progress_) = false;
+  if (error_code != boost::asio::error::operation_aborted) {
+    RAY_CHECK(redis_async_context_) << "redis_async_context_ must not be NULL";
+    {
+      // `redisAsyncHandleRead` and `redisAsyncHandleWrite` will mutate
+      // `redis_async_context_`, use a lock to protect it.
+      const std::lock_guard lock(mutex_);
+      write ? redisAsyncHandleWrite(redis_async_context_.get())
+            : redisAsyncHandleRead(redis_async_context_.get());
+    }
+  }
+
+  if (error_code == boost::asio::error::would_block) {
+    Operate();
+  }
+}
+
+void RedisAsyncContext::AddRead() {
+  // Because redis commands are non-thread safe, dispatch the operation to backend thread.
+  io_service_.dispatch(
+      [this] {
+        read_requested_ = true;
+        Operate();
+      },
+      "RedisAsyncContext.addRead");
+}
+
+void RedisAsyncContext::AddWrite() {
+  // Because redis commands are non-thread safe, dispatch the operation to backend thread.
+  io_service_.dispatch(
+      [this] {
+        write_requested_ = true;
+        Operate();
+      },
+      "RedisAsyncContext.addWrite");
+}
 
+void RedisAsyncContext::DelRead() { read_requested_ = false; }
+
+void RedisAsyncContext::DelWrite() { write_requested_ = false; }
+
+void RedisAsyncContext::Cleanup() {
```

Comment:
This used to be empty, is it intentional change?

Reply:
Reference from adapters examples in hiredis.

https://github.com/redis/hiredis/blob/8d8703ee6149ffd3b38cebedd71ad8d1fac63cc7/adapters/qt.h#L120-L123